text/0000-impl-trait-type-alias.md

```diff
@@ -0,0 +1,537 @@
+- Feature Name: impl-trait-type-alias
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add the ability to create type aliases for `impl Trait` types,
+and support `impl Trait` in `let`, `const`, and `static` declarations.
+
+```rust
+// `impl Trait` type alias:
+type Adder = impl Fn(usize) -> usize;
+fn adder(a: usize) -> Adder {
+    |b| a + b
+}
+
+// `impl Trait` type aliases in associated type position:
+struct MyType;
+impl Iterator for MyType {
+    type Item = impl Debug;
+    fn next(&mut self) -> Option<Self::Item> {
+        Some("Another item!")
+    }
+}
+
+// `impl Trait` in `let`, `const`, and `static`:
+
+const ADD_ONE: impl Fn(usize) -> usize = |x| x + 1;
+
+static MAYBE_PRINT: Option<impl Fn(usize)> = Some(|x| println!("{}", x));
+
+fn my_func() {
+    let iter: impl Iterator<Item = i32> = (0..5).map(|x| x * 5);
+    ...
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+This RFC proposes two expansions to Rust's `impl Trait` feature.
+`impl Trait`, first introduced in [RFC 1522][1522], allows functions to return
+types which implement a given trait, but whose concrete type remains anonymous.
+`impl Trait` was expanded upon in [RFC 1951][1951], which added `impl Trait` to
+argument position and resolved questions around syntax and parameter scoping.
+In its current form, the feature makes it possible for functions to return
+unnameable or complex types such as closures and iterator combinators.
+`impl Trait` also allows library authors to hide the concrete type returned by
+a function, making it possible to change the return type later on.
+
+However, the current feature has some severe limitations.
+Right now, it isn't possible to return an `impl Trait` type from a trait
+implementation. This is a huge restriction which this RFC fixes by making
+it possible to create a type alias for an `impl Trait` type using the syntax
+`type Foo = impl Trait;`. The type alias syntax also makes it possible to
+guarantee that two `impl Trait` types are the same:
+
+```rust
+// `impl Trait` in traits:
+struct MyStruct;
+impl Iterator for MyStruct {
+
+    // Here we can declare an associated type whose concrete type is hidden
+    // to other modules.
+    //
+    // External users only know that `Item` implements the `Debug` trait.
+    type Item = impl Debug;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        Some("hello")
+    }
+}
+```
+
+`impl Trait` type aliases allow us to declare multiple items which refer to
+the same `impl Trait` type:
+
+```rust
+// Type `Foo` refers to a type that implements the `Debug` trait.
+// The concrete type to which `Foo` refers is inferred from this module,
+// and this concrete type is hidden from outer modules (but not submodules).
+pub type Foo: impl Debug;
```

Comment:
Yes-- same for the other one. Thanks for catching that. I'll fix it as soon as I get to a computer.